### PR TITLE
init freezed_lint

### PR DESCRIPTION
**Pull request by SunlightBro**
#704
- freezed_missing_mixin

### CODE REVIEW

File: packages/freezed_lint/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Lints for [freezed].\
+This package does nothing without [freezed].
```
Comment (rrousselGit) [inline packages/freezed_lint/README.md:2]: Itd need a nice Readme :)

**Comment by rrousselGit** [comment]
Pretty cool!

**Comment by rrousselGit** [comment]
How's your experience with custom_lint so far?

**Comment by SunlightBro** [comment]
I came across one weird [bug #68](https://github.com/invertase/dart_custom_lint/issues/68) when applying the quick fix that the analyzer crashes.
I submitted the **log** as well as my lint **code**, because I'm not 100% sure if maybe it's my fault.

**Comment by rrousselGit** [comment]
Yeah I'm aware of it

It's not quite a bug. It's simply an error I need to silence the error and retry the analysis

**Comment by SunlightBro** [comment]
I'm still reading up [riverpod_lint_flutter_test](https://github.com/rrousselGit/riverpod/tree/master/packages/riverpod_lint_flutter_test) on how to properly write test for it, never worked with golden tests before.

**Comment by rrousselGit** [comment]
The separate package for tests was done because there's a dependency on Flutter (for Flutter_riverpod & hooks)

This isn't needed for Freezed


Also for goldens, you can use "expect_lint". And have the CI run custom_lint

File: packages/freezed_lint/README.md (hunk 1 above)
Comment (SunlightBro) [reply packages/freezed_lint/README.md]: updated

**Comment by SunlightBro** [comment]
Ok it's ready to review.
And pls squash merge this in the end, I made so many stupid mistakes in the github action 🙈.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -36,8 +39,9 @@ jobs:
           channel: ${{ matrix.channel }}
           cache: ${{ matrix.channel == 'stable' }}
 
-      # It is executed separately
+      # Retain example for lint golden test
       - name: Removing example folder
+        if: matrix.package != 'packages/freezed_lint'
```
Comment (rrousselGit) [inline .github/workflows/build.yml:44]:
```suggestion
      # It is executed separately
      - name: Removing example folder
        # Retain example for lint golden test
        if: matrix.package != 'packages/freezed_lint'
```

File: packages/freezed_lint/example/lib/missing_private_empty_ctor.dart [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import 'package:freezed_annotation/freezed_annotation.dart';
+
+part 'missing_private_empty_ctor.freezed.dart';
+
+@freezed
+// expect_lint: freezed_missing_private_empty_constructor
+class RequiresPrivateCtor with _$RequiresPrivateCtor {
+  const factory RequiresPrivateCtor() = _RequiresPrivateCtor;
+
+  void method() {
```
Comment (rrousselGit) [inline packages/freezed_lint/example/lib/missing_private_empty_ctor.dart:10]: We'd need a few more scenarios:
- an extend/mixin
- getters

File: packages/freezed_lint/lib/src/missing_mixin.dart [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import 'package:custom_lint_builder/custom_lint_builder.dart';
+import 'package:freezed_lint/src/tools/dart_type_ext.dart';
+import 'package:freezed_lint/src/tools/freezed_annotation_checker.dart';
+
+class MissingMixin extends DartLintRule {
+  MissingMixin() : super(code: _code);
+
+  static const _code = LintCode(
+    name: 'freezed_missing_mixin',
+    problemMessage: 'Freezed class should mixin with _\$<className>',
```
Comment (rrousselGit) [inline packages/freezed_lint/lib/src/missing_mixin.dart:12]: You can add variables in the code with `{0}`

Then do:

```dart
reporter.reportError(..., ['Hello world'])
```

File: packages/freezed_lint/lib/src/missing_mixin.dart [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      final element = node.declaredElement;
+      if (element == null) return;
+      final annotation = freezedAnnotationChecker.hasAnnotationOfExact(element);
+      if (!annotation) return;
+      final mixins = element.mixins;
+      final name = '_\$${element.name}';
+      if (mixins.any((e) => e.hasName(name))) return;
+      final src = element.source.contents.data;
+      if (src.contains(name)) return;
+      reporter.reportErrorForElement(code, element);
```
Comment (rrousselGit) [inline packages/freezed_lint/lib/src/missing_mixin.dart:32]:
```suggestion
      final annotation = freezedAnnotationChecker.hasAnnotationOfExact(element);
      if (!annotation) return;

      final mixins = element.mixins;
      final name = '_\$${element.name}';
      if (mixins.any((e) => e.hasName(name))) return;

      final src = element.source.contents.data;
      if (src.contains(name)) return;

      reporter.reportErrorForElement(code, element);
```

File: packages/freezed_lint/lib/src/missing_mixin.dart [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  @override
+  void run(
+    CustomLintResolver resolver,
+    ErrorReporter reporter,
+    CustomLintContext context,
+  ) {
+    context.registry.addClassDeclaration((node) {
+      final element = node.declaredElement;
+      if (element == null) return;
+      final annotation = freezedAnnotationChecker.hasAnnotationOfExact(element);
```
Comment (rrousselGit) [inline packages/freezed_lint/lib/src/missing_mixin.dart:25]:
```suggestion

      final annotation = freezedAnnotationChecker.hasAnnotationOfExact(element);
```

File: packages/freezed_lint/lib/src/missing_mixin.dart [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        if (element == null) return;
+        final name = element.displayName;
+        if (element.mixins.isEmpty) {
+          builder.addSimpleInsertion(
+            element.nameOffset + element.nameLength,
+            ' with _\$$name',
+          );
+        } else {
+          builder.addSimpleInsertion(
+            element.nameOffset + element.nameLength + 5,
```
Comment (rrousselGit) [inline packages/freezed_lint/lib/src/missing_mixin.dart:70]: These offsets work if the class extends something

Comment (rrousselGit) [reply packages/freezed_lint/lib/src/missing_mixin.dart:70]: (same for the other case too)

Comment (rrousselGit) [reply packages/freezed_lint/lib/src/missing_mixin.dart:70]: Also what's that `+5`?

File: packages/freezed_lint/lib/src/missing_mixin.dart [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+import 'package:freezed_lint/src/tools/dart_type_ext.dart';
+import 'package:freezed_lint/src/tools/freezed_annotation_checker.dart';
+
+class MissingMixin extends DartLintRule {
+  MissingMixin() : super(code: _code);
+
+  static const _code = LintCode(
+    name: 'freezed_missing_mixin',
+    problemMessage: 'Freezed class should mixin with _\$<className>',
+    errorSeverity: ErrorSeverity.WARNING,
```
Comment (rrousselGit) [inline packages/freezed_lint/lib/src/missing_mixin.dart:13]: Let's use the default severity. It's not a critical issue

File: packages/freezed_lint/lib/src/missing_private_empty_ctor.dart [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+class MissingPrivateEmptyCtor extends DartLintRule {
+  MissingPrivateEmptyCtor() : super(code: _code);
+
+  static const _code = LintCode(
+    name: 'freezed_missing_private_empty_constructor',
+    problemMessage: 'Private empty constructor required',
+    correctionMessage:
+        'Freezed classes containing methods, fields or accessors,'
+        'require a const <className>._()',
+    errorSeverity: ErrorSeverity.WARNING,
```
Comment (rrousselGit) [inline packages/freezed_lint/lib/src/missing_private_empty_ctor.dart:15]: This probably should be an error
As the compiler will fail

File: packages/freezed_lint/lib/src/missing_mixin.dart (hunk 7 above)
Comment (SunlightBro) [reply packages/freezed_lint/lib/src/missing_mixin.dart:70]: length of `' with'`

File: packages/freezed_lint/lib/src/missing_private_empty_ctor.dart [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    context.registry.addClassDeclaration((node) {
+      final element = node.declaredElement;
+      if (element == null) return;
+      final annotation = freezedAnnotationChecker.hasAnnotationOfExact(element);
+      if (!annotation) return;
+      final methods = element.methods;
+      final fields = element.fields;
+      final accessors = element.accessors;
+      if (methods.isEmpty && fields.isEmpty && accessors.isEmpty) return;
+      final ctors = element.constructors.where((ctor) {
```
Comment (rrousselGit) [inline packages/freezed_lint/lib/src/missing_private_empty_ctor.dart:33]:
```suggestion
      final element = node.declaredElement;
      if (element == null) return;

      final annotation = freezedAnnotationChecker.hasAnnotationOfExact(element);
      if (!annotation) return;

      final methods = element.methods;
      final fields = element.fields;
      final accessors = element.accessors;
      if (methods.isEmpty && fields.isEmpty && accessors.isEmpty) return;

      final ctors = element.constructors.where((ctor) {
```

File: packages/freezed_lint/lib/src/missing_private_empty_ctor.dart [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+      if (methods.isEmpty && fields.isEmpty && accessors.isEmpty) return;
+
+      final ctors = element.constructors.where((ctor) {
+        return ctor.isConst &&
+            ctor.isPrivate &&
+            ctor.parameters.isEmpty &&
+            ctor.name == '_';
+      });
+      if (ctors.isNotEmpty) return;
+      reporter.reportErrorForElement(code, element);
```
Comment (rrousselGit) [inline packages/freezed_lint/lib/src/missing_private_empty_ctor.dart:43]:
```suggestion

      reporter.reportErrorForElement(code, element);
```

File: packages/freezed_lint/lib/src/missing_private_empty_ctor.dart [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    CustomLintResolver resolver,
+    ChangeReporter reporter,
+    CustomLintContext context,
+    AnalysisError analysisError,
+    List<AnalysisError> others,
+  ) {
+    context.registry.addClassDeclaration((node) {
+      if (!analysisError.sourceRange.intersects(node.sourceRange)) return;
+      final changeBuilder = reporter.createChangeBuilder(
+        message: 'Add const ${node.name}._();',
```
Comment (rrousselGit) [inline packages/freezed_lint/lib/src/missing_private_empty_ctor.dart:64]: It should be const only if one of the Freezed constructor is const too

It wouldn't make sense for it to be `const` on an `@unfreezed` for instance

File: packages/freezed_lint/lib/src/missing_private_empty_ctor.dart [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        priority: 2,
+      );
+      changeBuilder.addDartFileEdit((builder) {
+        final element = node.declaredElement;
+        if (element == null) return;
+        final name = element.displayName;
+        final nextLine =
+            resolver.lineInfo.getOffsetOfLineAfter(element.nameOffset);
```
Comment (rrousselGit) [inline packages/freezed_lint/lib/src/missing_private_empty_ctor.dart:72]: You can't rely on the name here.

Someone may write:

```dart
class Foo
{
  
}
```

Comment (rrousselGit) [reply packages/freezed_lint/lib/src/missing_private_empty_ctor.dart:72]: There's also the possibility of:

```dart
class Foo
  with
    A,
    B,
    C {
    
}
```

which is actually something dartfmt does

File: packages/freezed_lint/test/ensure_build_test.dart [hunk 14]
```diff
@@ -0,0 +1,10 @@
+import 'package:build_verify/build_verify.dart';
+import 'package:test/test.dart';
+
+void main() {
+  test('ensure_build', () {
+    expectBuildClean(
```
Comment (rrousselGit) [inline packages/freezed_lint/test/ensure_build_test.dart:6]: Fixes aren't tested. Although I'll admit I haven't thought too much about how to test them.

File: packages/freezed_lint/lib/src/missing_mixin.dart (hunk 7 above)
Comment (rrousselGit) [reply packages/freezed_lint/lib/src/missing_mixin.dart:70]: There could be any number of space between the name and `with` or an `extends` clause

You should be able using the AST to find the exact location of the `with` token

**Comment by rrousselGit** [comment]
LGTM, nice job :)

**Pull request merged**